### PR TITLE
additional ports support

### PR DESCRIPTION
**Pull request by vhvb1989**
fix: https://github.com/Azure/azure-dev/issues/3267

### CODE REVIEW

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		return nil, nil
+	}
+
+	if err := validateInput(bindings); err != nil {
+		return nil, err
+	}
+
+	endpointByContainerPort := make(map[string][]*Binding)
+	for _, binding := range bindings {
+		bindingKey := "default" // default is for those with no container port
```
Comment (weikanglim) [inline cli/azd/pkg/apphost/aca_ingress.go:19]: I don't know if it's acceptable to just use `0` here as an "automatically assigned port".

**Review by weikanglim — COMMENTED** [review]
Overall looks good. Just brought up some questions if we want to simplify `buildAcaIngress` -- it's 80+ lines long, and I did find it a little hard to parse through.

File: cli/azd/pkg/apphost/aca_ingress.go (hunk 1 above)
Comment (vhvb1989) [reply cli/azd/pkg/apphost/aca_ingress.go:19]: It won't use 0, 
`http-only` endpoints can omit the TargetPort and they just get the `defaultPort` argument (either 80 or 8080)

File: cli/azd/pkg/apphost/aca_ingress_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+			TargetPort: to.Ptr(8084),
+			External:   false,
+			Scheme:     acaIngressSchemaHttp,
+			Transport:  acaIngressTransportHttp2,
+		},
+	}
+	ingress, err = buildAcaIngress(bindings, 8080)
+	assert.Error(t, err)
+	assert.EqualError(t, err, "More than 5 additional ports are not supported. "+
+		"See https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview#tcp for more details.")
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress_test.go:135]: Can you remove the `en-us` part of this URL?  Azure will redirect to the correct location based on the user's language (other projects in the larger azure-sdk org have some linters that run over markdown files and warn about this, which is why I know this trivia)

File: cli/azd/pkg/apphost/aca_ingress_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+import (
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBuildAcaIngress(t *testing.T) {
+	// Test case 1: Empty bindings
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress_test.go:11]: Possible that the indention hit is not worth it, but if you had used `t.Run` to make these all subtests, I think we could have both used `t.Parallel()` and it would have been easier to run each test under the debugger as an atomic thing.

File: cli/azd/pkg/apphost/generate.go [hunk 4]
```diff
@@ -1042,7 +993,7 @@ func (b *infraGenerator) Compile() error {
 			KeyVaultSecrets: make(map[string]string),
 		}
 
-		ingress, err := buildIngress(docker.Bindings)
+		ingress, err := buildAcaIngress(docker.Bindings, 80)
```
Comment (ellismg) [inline cli/azd/pkg/apphost/generate.go:996]: Just checking my understanding - while we pass `80` here as a default port, for dockerfiles the manifest is always going to tell us the ingress port so we won't end up using this, but the contract of `buildAcaIngress` forces us to pass something, so we pass the default HTTP port, but we know it won't be used?

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+type endpointGroupProperties struct {
+	port     int
+	external bool
+	httpOnly bool
+	hasHttp2 bool
+}
+
+const (
+	acaIngressSchemaTcp      string = "tcp"
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress.go:145]: Sort of a NIT, but you might consider defining three types like:

```golang
type acaIngressScheme string
type acaIngressTransport string
type acaIngressProtocol string
```

And these use these types here (and in the model) types to get a bit of type safety.  I'm just a little worried about us mixing these constants together which feels a little easier when they are all typed as string.

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	for containerPort, bindings := range endpointByContainerPort {
+		props := &endpointGroupProperties{httpOnly: true}
+		for _, binding := range bindings {
+			if binding.TargetPort != nil {
+				props.port = *binding.TargetPort
+			}
+			if binding.External {
+				props.external = true
+			}
+			if binding.Scheme != acaIngressSchemaHttp && binding.Scheme != acaIngressSchemaHttps {
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress.go:36]: Just wondering - is it an error for one of the bindings (for the same port) to have an scheme of `tcp` while another has `http` or `https`?  I think that it is an in practice such a manifest would never exist, but I am wondering if we wanted to have some validation about that?

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+		}
+	}
+
+	additionalPortsCount := len(endpointByContainerPort)
+	if ingress != "" {
+		additionalPortsCount--
+	}
+	if additionalPortsCount > 5 {
+		return nil, fmt.Errorf("More than 5 additional ports are not supported. " +
+			"See https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview#tcp for more details.")
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress.go:93]: Since I read the test code first, I commented on the `en-us` bit there, but here it is in the product, if this helps you find it faster and remove it :-)

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
+package apphost
+
+import (
+	"fmt"
+)
+
+// buildAcaIngress builds the Azure Container Apps ingress configuration from the provided bindings.
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress.go:7]: I feel like some high level description of the algorithm here might be useful - We end up doing multiple levels of grouping here, it might be helpful to orient the reader about what's about to happen as we try to merge multiple bindings together to build the ACA ingress configuration.

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+		if props.hasHttp2 {
+			finalIngress.Transport = acaIngressTransportHttp2
+		}
+		if props.hasHttp2 || !props.external {
+			finalIngress.AllowInsecure = true
+		}
+
+	} else {
+		port := 0
+		for groupKey := range endpointByContainerPort {
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress.go:113]: Because enumerating a go map does not have a fixed order, in practice this is just picking a random entry to use as the default ingress rule and then the rest will be used as additional rules.  Is that what we want?  I'm not sure.

Comment (ellismg) [reply cli/azd/pkg/apphost/aca_ingress.go:113]: I can image that it doesn't matter - but a helpful comment about it why it's okay to just pick one at random and then treat the rest as additional ports would be good?  I feel like in practice this might only be able to happen if every endpoint is TCP only (because otherwise we'd try to make one of the HTTP/HTTPs the default endpoint so we could use the ACA ingress controller?)

**Review by ellismg — COMMENTED** [review]
I do think that this is pretty close (and the detailed tests give me a better understanding of what's going on and confidence this is right, which is great!) - but I will admit that I have sort of forgotten the algorithm that we discussed with David and his C# code sample, so I can't say for sure that I understand exactly what's going on here and can confirm it was correct.  I made a few comments about the actual merging of values with things that stuck out or confused me, but I feel like this sort of thing is the place where a few helpful comments as we go along about what we are doing (and what invariants we've established) might be helpful.

File: cli/azd/pkg/apphost/aca_ingress.go (hunk 5 above)
Comment (vhvb1989) [reply cli/azd/pkg/apphost/aca_ingress.go:145]: I started like that, but then I need to do something like 

```go
if acaIngressScheme(string) == acaIngressSchemeHttp {
}
```

That's why I ended up just using strings. Under the assumption, as well, that this is only inside the package and not exported

File: cli/azd/pkg/apphost/aca_ingress_test.go (hunk 3 above)
Comment (vhvb1989) [reply cli/azd/pkg/apphost/aca_ingress_test.go:11]: I started from an auto-generated copilot template :)

**Comment by davidfowl** [comment]
After testing this, we need to relax some of the rules around how we pick the http ingress. If there are multiple HTTP internal endpoints, the first one becomes the ingress and the rest are additional ports.

**Comment by davidfowl** [comment]
I made that change here https://github.com/dotnet/aspire/pull/2965/commits/9077747e27fc0f6205f728fcb1e2a010ca2d3bc7

**Comment by davidfowl** [comment]
From our meeting, instead of throwing because of the 5 endpoint limitation, we warn but let the deployment go through. This means we don't need to update azd if a customer gets that restriction relaxed.

File: cli/azd/pkg/apphost/generate.go (hunk 4 above)
Comment (vhvb1989) [reply cli/azd/pkg/apphost/generate.go:1006]: Yes,  `buildAcaIngress` doesn't know if it is building for a docker file, container or a project.
All it knows is that there might be a case where there are http bindings without explicitly mentioning a targetPort. For that only case, the method is asking what port to use as default. And it is not baked inside the method because we use 80 or 8080

File: cli/azd/pkg/apphost/aca_ingress.go (hunk 6 above)
Comment (davidfowl) [reply cli/azd/pkg/apphost/aca_ingress.go:113]: Yes it is, I added an error case for this. We can always allow it later but detection isn't bad here.

File: cli/azd/pkg/apphost/manifest.go [hunk 10]
```diff
@@ -103,11 +103,35 @@ type Reference struct {
...
+	External   bool   `json:"external"`
+}
+
+// WithIndexKey keeps the insertion order of the bindings.
+type WithIndexKey struct {
+	string
+	Index int
+}
+
+var withIndexCounter = 0
```
Comment (ellismg) [inline cli/azd/pkg/apphost/manifest.go:119]: Take a look at https://github.com/ellismg/azure-dev/commit/7baa890f15b744ab8ca8d29c33bb4899c9df621e which introduces a `bindingsMap` type which customizes its UnmarshalJSON to also retain a slice with the ordered keys from the object. 

I would rather use this type instead of `map[WithIndexKey]*Binding`.

File: cli/azd/pkg/apphost/aca_ingress.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+		for _, httpGroup := range httpOnlyGroups {
+			if endpointByTargetPortProperties[httpGroup].external {
+				externalHttpOnly = append(externalHttpOnly, httpGroup)
+			}
+		}
+
+		if len(externalHttpOnly) == 1 {
+			ingress = externalHttpOnly[0]
+		} else if len(httpOnlyGroups) > 1 {
+			minIndex := 0
```
Comment (ellismg) [inline cli/azd/pkg/apphost/aca_ingress.go:172]: The logic will have to change here if we adopt my idea where we have a slice of ordered keys - but I think it should be possible to do?

**Review by ellismg — COMMENTED** [review]
Thanks for the improvements to the readability of the core binding algorithm.  Overall looks good - but let's adopt something like what I outlined to track the order of the keys instead of a global written to by UnmarshalText.  Once that's done, I think this will be good!

File: cli/azd/pkg/apphost/manifest.go (hunk 10 above)
Comment (vhvb1989) [reply cli/azd/pkg/apphost/manifest.go]: I used your prototype to create `type WithOrder[T any] struct` (here: https://github.com/Azure/azure-dev/pull/3612/files#diff-48bd9ed57e1e9d7efb4412f0530ff8c7b10653a8691714a00bd073ee961a1b8fR10)

This type would help us to re-use the `map with-order` again if we ever need it

**Review by weikanglim — APPROVED** [review]
